### PR TITLE
logits_to_keep bool mask for all ForXXX classes, expose last_hidden_state, fix negative output_hidden_states indices

### PR DESCRIPTION
**Pull request by jp1924**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48825) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48825)

Refs #48784

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @zucchini-nlp since you commented on the issue!

**Comment by VimalN2005** [comment]
Hi @jp1924 @zucchini-nlp @vasqu,

I verified this issue locally and added:
1. **Full regression test coverage** in `tests/models/gemma4/test_modeling_gemma4.py` (`test_logits_to_keep_bool_mask_and_last_hidden_state`) verifying non-contiguous bool tensor masking and `last_hidden_state` exposure.
2. **Negative layer index resolution** in `src/transformers/utils/output_capturing.py` so that requests like `output_hidden_states=[-1]` (e.g. from candidate generator) resolve properly against hook counts without caching intermediate layers.

I have all tests passing locally. I have pushed the branch to my fork [here](https://github.com/VimalN2005/transformers/tree/fix-gemma4-logits-mask-and-last-hidden-state). 

If you'd like, I can open a PR against @jp1924's branch to fold the tests and hook improvements in, or open a standalone PR!

**Comment by zucchini-nlp** [comment]
Ping us when you are ready for review @jp1924 , atm the only comment is to not add a new `last_hidden_states` field. As discussed we can land `output_hidden_states_indices` to get a more scalable feat

**Comment by jp1924** [comment]
@zucchini-nlp @ArthurZucker

PR is open.
Patching ended up touching around 300 files.
Still reviewing it myself - most diffs are the same +9/-2 pattern, so I prioritized reviewing the files that deviate from that.
Tests and check-repo passed on my end, but the scope is large enough that I'm still going back over it.

**Comment by jp1924** [comment]
@VimalN2005

Thanks for looking into this.
I'd already finished the work though.
Hope we get a chance to collaborate sometime.

**Comment by jp1924** [comment]
https://github.com/huggingface/transformers/issues/48784#issuecomment-5696954890
https://github.com/huggingface/transformers/issues/48784#issuecomment-5696536700

This is something I've been thinking about while working on this feature.

As mentioned in the comments linked above, logits_to_keep looks like it was originally built for use inside generate(). int and 1D IndexTensor were enough for that, and this patch adds 2D Tensor support on top. But having this many branches on logits_to_keep isn't great. Eventually unifying generate() onto 2D Tensor too would be better for long-term code health. That's a much bigger change than this PR's scope though, so it's staying as a patch for now.

Not sure how you'll want to decide this, but if possible, adding a deprecation warning down the line to formally sunset int/1D support seems worth doing

File: src/transformers/models/afmoe/modeling_afmoe.py [hunk 1]
```diff
@@ -660,8 +660,15 @@ def forward(
...
         hidden_states = outputs.last_hidden_state
-        slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
-        logits = self.lm_head(hidden_states[:, slice_indices, :])
+        if isinstance(logits_to_keep, int):
+            slice_indices = slice(-logits_to_keep, None)
+            hidden_states = hidden_states[:, slice_indices, :]
+        elif logits_to_keep.dtype == torch.bool:
+            hidden_states = hidden_states[logits_to_keep]
+        else:
+            hidden_states = hidden_states[:, logits_to_keep, :]
```
Comment (vasqu) [inline src/transformers/models/afmoe/modeling_afmoe.py:669]:
```suggestion
        if logits_to_keep is not None:
            if isinstance(logits_to_keep, int):
                logits_to_keep = torch.arange(
                    hidden_states.shape[1] - logits_to_keep,
                    hidden_states.shape[1],
                    device=hidden_states.device,
                )
        
            indices = logits_to_keep[None] if logits_to_keep.ndim == 1 else logits_to_keep
            hidden_states = torch.take_along_dim(hidden_states, indices[..., None], dim=1)
```
Just thinking of way to make this a bit more elegant

I'd expect that we allow 
1. int (last x tensors)
2. bool / long / int tensor that either goes on the seq len or more specialized on the batch x seq

I really doubt that it makes sense to deprecate one variation over the other. There is a lot of history and it seems unnecessarily breaking. What we essentially extend with this PR is that you can additionally have a 2d tensor instead of 1d types only (1d tensor, int). The dtype is essentially less important tbh as long as it acts as indexing

What is important is that we update the description of logits_to_keep in autodocstring.py

File: src/transformers/utils/output_capturing.py
Comment (vasqu) [inline src/transformers/utils/output_capturing.py:1]: Let's move this to a separate PR pls, it's 2 functionalities we want to add
1. logits to keep with 2d indexing
2. output capturing with OPTIONAL last hidden state

File: src/transformers/utils/output_capturing.py [hunk 2]
```diff
@@ -270,15 +276,21 @@ def wrapper(self, *args, **kwargs):
...
             ):
-                collected_outputs["_hidden_states_layers"] = set(recordable_keys["output_hidden_states"])
-            # Make sure hooks are installed if we need to collect outputs
-            if len(collected_outputs) > 0:
-                maybe_install_capturing_hooks(self)
+                num_hidden_states_layers = getattr(self, "_output_capturing_hook_counts", {}).get("hidden_states", 0)
+                hidden_states_layers = set()
+                for idx in recordable_keys["output_hidden_states"]:
+                    hidden_states_layers.add(idx if idx >= 0 else num_hidden_states_layers + idx)
+                collected_outputs["_hidden_states_layers"] = hidden_states_layers
```
Comment (vasqu) [inline src/transformers/utils/output_capturing.py:293]: This feels all very implicit, we should make it explicitly to be requested by the user so there is no magic happening

**Review by vasqu — COMMENTED** [review]
Some initial comments
